packages/flutter/lib/src/widgets/raster_widget.dart

```diff
@@ -0,0 +1,401 @@
+        _onRasterValueChanged();
+      }
+    }
+  }
+
+  /// The delegate used to draw the image representing the child.
+  RasterWidgetDelegate get delegate => _delegate;
+  RasterWidgetDelegate _delegate;
+  set delegate(RasterWidgetDelegate value) {
+    if (value == delegate) {
+      return;
+    }
+    delegate.removeListener(markNeedsPaint);
+    final RasterWidgetDelegate oldDelegate = _delegate;
+    _delegate = value;
+    if (attached) {
+      delegate.addListener(markNeedsPaint);
+      if (delegate.shouldRepaint(oldDelegate)) {
+        markNeedsPaint();
+      }
+    }
+  }
+
+  /// A fallback delegate which is used if the child layers contains a platform view.
+  RasterWidgetFallbackDelegate? get fallback => _fallback;
+  RasterWidgetFallbackDelegate? _fallback;
+  set fallback(RasterWidgetFallbackDelegate? value) {
+    if (value == fallback) {
+      return;
+    }
+    _fallback = value;
+    markNeedsPaint();
+  }
+
+  /// Whether the raster widget will ignore platform views in child layers.
+  bool get ignorePlatformViews => _ignorePlatformViews;
+  bool _ignorePlatformViews;
+  set ignorePlatformViews(bool value) {
+    if (value == ignorePlatformViews) {
+      return;
+    }
+    _ignorePlatformViews = value;
+    if (ignorePlatformViews) {
+      _hitPlatformView = false;
+    }
+    markNeedsPaint();
+  }
+
+  ui.Image? _childRaster;
+
+  @override
+  void attach(covariant PipelineOwner owner) {
+    delegate.addListener(markNeedsPaint);
+    controller.addListener(_onRasterValueChanged);
+    super.attach(owner);
+  }
+
+  @override
+  void detach() {
+    delegate.removeListener(markNeedsPaint);
+    controller.removeListener(_onRasterValueChanged);
+    _childRaster?.dispose();
+    _childRaster = null;
+    super.detach();
+  }
+
+  @override
+  void dispose() {
+    delegate.removeListener(markNeedsPaint);
+    controller.removeListener(_onRasterValueChanged);
+    _childRaster?.dispose();
+    _childRaster = null;
+    super.dispose();
+  }
+
+  void _onRasterValueChanged() {
+    _childRaster?.dispose();
+    _childRaster = null;
+    markNeedsPaint();
+  }
+
+  bool _hitPlatformView = false;
+
+  // Paint [child] with this painting context, then convert to a raster and detach all
+  // children from this layer.
+  ui.Image? _paintAndDetachToImage(PaintingContext originalContext) {
+    final OffsetLayer offsetLayer = OffsetLayer();
+    final PaintingContext context = PaintingContext(offsetLayer, Offset.zero & size);
+    super.paint(context, Offset.zero);
+    // This ignore is here because this method is protected by the `PaintingContext`. Adding a new
+    // method that performs the work of `_paintAndDetachToImage` would avoid the need for this, but
+    // that would conflict with our goals of minimizing painting context.
+    // ignore: invalid_use_of_protected_member
+    context.stopRecordingIfNeeded();
+    if (!ignorePlatformViews && !offsetLayer.supportsRasterization()) {
```

Comment:
THere's an implicit walk of the layer tree here. This doesn't change the perf characteristics in an undesired way?

Reply:
Because we only walk the layer tree once per raster widget (and because we need to walk it anyway to render it) - it is fairly cheap.